src/devices/honeywell_cm921.c

```diff
@@ -0,0 +1,449 @@
+    for (size_t i = 0; i < sizeof(device_map) / sizeof(dev_map_entry_t); i++)
+        if (device_map[i].t == dev_type)
+            dev_name = device_map[i].s;
+
+    char *res = malloc(sizeof(char) * (6 + strlen(dev_name) + 1));
+    sprintf(res, "%3s:%06d", dev_name, (device_id[0] & 0x03) << 16 | (device_id[1] << 8) | device_id[2]);
+    return res;
+}
+
+data_t *decode_device_ids(const message_t *msg, data_t *data, int style) {
+    char *ds = calloc(11, msg->num_device_ids);
+
+    for (unsigned i = 0; i < msg->num_device_ids; i++) {
+        if (i != 0) strcat(ds, " ");
+
+        char *did = NULL;
+        if (style == 0)
+            did = decode_device_id(msg->device_id[i]);
+        else {
+            did = malloc(sizeof(char) * 7);
+            for (unsigned j = 0; j < 3; j++)
+                sprintf(did + 2*j, "%02x", msg->device_id[i][j]);
+        }
+        strcat(ds, did);
+        free(did);
+    }
+
+    return data_append(data, "Device IDs", "", DATA_STRING, ds, NULL);
+}
+
+#define UNKNOWN_IF(C) if (C) { return data_append(data, "unknown", "", DATA_FORMAT, "%04x", DATA_INT, msg->command, NULL); }
+
+data_t *interpret_message(const message_t *msg, data_t *data)
+{
+    /* Sources of inspiration:
+       https://github.com/Evsdd/The-Evohome-Protocol/wiki
+       https://www.domoticaforum.eu/viewtopic.php?f=7&t=5806&start=30
+       (specifically https://www.domoticaforum.eu/download/file.php?id=1396) */
+
+    data = decode_device_ids(msg, data, 1);
+
+    data_t *r = data;
+    switch(msg->command) {
+        case 0x1030: {
+            UNKNOWN_IF(msg->payload_length != 16);
+            data = data_append(data, "zone_idx", "", DATA_FORMAT, "%02x", DATA_INT, msg->payload[0], NULL);
+            for (unsigned i = 0; i < 5; i++) { // order fixed?
+                const uint8_t *p = &msg->payload[1 + 3*i];
+                // *(p+1) == 0x01 always?
+                int value = *(p+2);
+                switch (*p) {
+                    case 0xC8: data = data_append(data, "max_flow_temp", "", DATA_INT, value, NULL); break;
+                    case 0xC9: data = data_append(data, "pump_run_time", "", DATA_INT, value, NULL); break;
+                    case 0xCA: data = data_append(data, "actuator_run_time", "", DATA_INT, value, NULL); break;
+                    case 0xCB: data = data_append(data, "min_flow_temp", "", DATA_INT, value, NULL); break;
+                    case 0xCC: /* Unknown, always 0x01? */ break;
+                    default: printf("Unknown parameter to 0x1030: %x02d=%04d\n", *p, value);
+                }
+            }
+            break;
+        }
+        case 0x313F: {
+            UNKNOWN_IF(msg->payload_length != 1 && msg->payload_length != 9);
+            switch (msg->payload_length) {
+                case 1:
+                    data = data_append(data, "time_request", "", DATA_INT, msg->payload[0], NULL); break;
+                case 9: {
+                    X313F* pl = (X313F*)msg->payload;
```

Comment:
Not fond of this construct.

Reply:
Neither am I, it doesn't work in this context. I got rid of it.